### PR TITLE
[dotnet] Added tests, fixed bugs

### PR DESCRIPTION
**Pull request by stephen-hawley**
Stunt testing!

What happens is this: we write a string of C# to a file. This C# acts as a mock of the ObjCRuntime namespace it also includes a one or more classes that act internally like NSObject descendants at least as far as the `class_ptr` field is concerned.

The test compiles this code, runs the rewriter, then executes the final code collecting the output.
The tests include the following:

1. tool not run, code works as expected
2. tool run, `ClassHandle` now returns a different value
3. tool run, a class not listed for changing stays the same
4. tool run on a more authentic cctor, the cctor gets removed entirely

The code to run a process and collect its output was taken from the binding-tools-for-swift tests.

All tests pass.

### CODE REVIEW

File: tools/class-redirector/class-redirector-tests/Compiler.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	static string GetTempDirectory ()
+	{
+		var unique = Interlocked.Increment (ref counter).ToString ();
+		var path = Path.Combine (root, unique);
+		Directory.CreateDirectory (path);
+		return path;
+	}
+
```
Comment (tj-devel709) [inline tools/class-redirector/class-redirector-tests/Compiler.cs:43]:
```suggestion
```
nit: reformatter may grab this :)

File: tools/class-redirector/class-redirector-tests/Compiler.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	{
+		var info = new ProcessStartInfo (path, args);
+		info.UseShellExecute = false;
+		info.RedirectStandardInput = false;
+		info.RedirectStandardOutput = true;
+		info.RedirectStandardError = true;
+		if (workingDirectory != null)
+			info.WorkingDirectory = workingDirectory;
+
+		if (output == null)
```
Comment (tj-devel709) [inline tools/class-redirector/class-redirector-tests/Compiler.cs:88]:
```suggestion
		if (output is null)
```

File: tools/class-redirector/class-redirector-tests/Compiler.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	public static int RunCommand (string path, string args, IDictionary<string, string>? env = null, StringBuilder? output = null, bool verbose = false, string? workingDirectory = null)
+	{
+		var info = new ProcessStartInfo (path, args);
+		info.UseShellExecute = false;
+		info.RedirectStandardInput = false;
+		info.RedirectStandardOutput = true;
+		info.RedirectStandardError = true;
+		if (workingDirectory != null)
```
Comment (tj-devel709) [inline tools/class-redirector/class-redirector-tests/Compiler.cs:85]:
```suggestion
		if (workingDirectory is not null)
```

File: tools/class-redirector/class-redirector-tests/Compiler.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			info.WorkingDirectory = workingDirectory;
+
+		if (output == null)
+			output = new StringBuilder ();
+
+		if (env != null) {
```
Comment (tj-devel709) [inline tools/class-redirector/class-redirector-tests/Compiler.cs:91]:
```suggestion
		if (env is not null) {
```

File: tools/class-redirector/class-redirector-tests/Compiler.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			foreach (var kvp in env) {
+				if (kvp.Value == null) {
```
Comment (tj-devel709) [inline tools/class-redirector/class-redirector-tests/Compiler.cs:93]:
```suggestion
				if (kvp.Value is null) {
```

**Review by tj-devel709 — COMMENTED** [review]
Few very small things, but otherwise LGTM :)

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**